docs/compilers/CSharp/Nullability Public API Design Notes.md

```diff
@@ -16,17 +16,22 @@ This includes features like intellisense.
 
 ## General Concepts
 
-Nullability information is exposed publicly as an enum, `Nullability`.
-There are 5 possible values:
-
-* `NotApplicable` - This is used when nullability information is requested about something that does not have a nullability, such as a statement, directive, or other similar constructs.
-* `NotAnalyzed` - This is used to fulfill priority #2.
-If the user requests information while asking for nullability to not be calculated, we'll use this to stand in for nullability.
-* `Unknown` - This is used for legacy C# code that does not know about nullability.
-This can appear in modern C# code that calls unannotated C#, or in certain generic variance scenarios where a common nullability cannot be inferred from the supplied arguments.
-For use in this document unknown is represented with the `~` character, but this character does not appear in source.
-* `NotNull` - This is used when we know the nullability is not null.
-* `MayBeNull` - This is used when we know the nullability may be null.
+Nullability information is exposed publicly as two enums, `NullableAnnotation` and `NullableFlowState`.
+
+`NullableAnnotation` represents lvalues. There are 4 possible values:
+
+* `Default` - This is used when nullability information is requested about something that does not have the concept of nullability, such as a statement, directive, or other similar constructs.
+If we decide to expose a version of `SemanticModel` that does not compute nullability information for certain perf-sensitive scenarios, all annotation information returned would be `Default`.
+* `Unknown` - This value was defined in a nullable-disabled context.
+* `NotAnnotated` - This value was defined in a nullable-enabled context, and was not annotated in source (or in type inference scenarios, was inferred to be a not annotated type).
+* `Annotated` - This value was defined in a nullable-enabled context, and was annotated in source (or in type inference scenarios, was inferred to be a annotated type).
```

Comment:
We can't make the names here `MaybeNull` and `NotNull`: that's not what they mean. They whether or not the source is annotated with a `?` or not. There's not a direct correspondence between these concepts: a `NotAnnotated` variable of type `T`, where `T` is an unconstrained type parameter, has a flow state of `MaybeNull` when used as an rvalue. However, you cannot assign a `MaybeNull` value to it. Unconstrained type parameters are going to be confusing for everyone, I think: users, library authors, the IDE, us...

Reply:
Ah, I think I see the mental breakdown here for me: I don't think of a parameter symbol, say, having an lvalue nullability: I read parameters far more than I ever write to them after all! And the the most common "write" to a parameter is really a method call...

Spitballing here: what if we had an enum (maybe this one, maybe something else) that has CanBeGivenNull, CannotBeGivenNull? It makes the lvalue-ness far more obvious.

If we didn't want to do that, I'd at least call AnnotatedAsNonNull and NotAnnotated, because still unclear what "annotated" otherwise means, _especially_ if/when you ever need another kind of annotation as future proofing.